File: framework/nexum-core/lib/core/services/packet_deserializer_service.dart
import 'package:nexum_core/exceptions/packet_deserialization_exception.dart';
import 'package:nexum_core/channel/friendly_buffer.dart';
import 'package:nexum_core/channel/packet.dart';

import '../../channel/packet_deserializer.dart';
import '../storages/packet_deserializer_registry.dart';

class PacketDeserializerService {
  static PacketDeserializerService get instance => _instance;
  static const PacketDeserializerService _instance = PacketDeserializerService._();

  const PacketDeserializerService._();

  static Packet deserializePacket(FriendlyBuffer friendlyBuffer) {
    final String identifier = friendlyBuffer.read("identifier");
    final String uuid = friendlyBuffer.read("packet_uuid");

    final PacketDeserializer ? packetDeserializer = PacketDeserializerRegistry.instance.get(identifier);

    if(packetDeserializer == null) {
      throw PacketDeserializationException(
        "Não foi possivel encontrar o deserializador para o pacote: $identifier"
      );
    }


    final Packet packet = packetDeserializer.deserialize(friendlyBuffer);
    packet.uuid = uuid;

    return packet;
  }
}

File: engine/nexum-core/src/main/java/io/nexum/core/services/PacketDeserializerService.java
package io.nexum.core.services;

import io.nexum.core.storages.PacketDeserializerRegistry;
import io.nexum.exceptions.PacketDeserializationException;
import io.nexum.channel.FriendlyBuffer;
import io.nexum.channel.Packet;
import io.nexum.channel.PacketDeserializer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.UUID;

public class PacketDeserializerService {
    private static @NotNull PacketDeserializerService instance = new PacketDeserializerService();

    private PacketDeserializerService() {

    }

    public <T extends Packet> @NotNull T deserializePacket(@NotNull FriendlyBuffer friendlyBuffer) {
        final String identifier = friendlyBuffer.read("identifier");
        final UUID uuid = UUID.fromString(friendlyBuffer.read("packet_uuid"));

        @Nullable
        final PacketDeserializer<Packet> packetDeserializer = PacketDeserializerRegistry.getInstance().get(identifier);

        if(packetDeserializer == null) throw new PacketDeserializationException(
                String.format("O desserializador de identificador %s não foi encontrado", identifier)
        );

        final Packet packet = packetDeserializer.deserialize(friendlyBuffer);

        packet.setUUID(uuid);
        @SuppressWarnings("unchecked")
        final T casted = (T) packet;

        return casted;
    }

    public static @NotNull PacketDeserializerService getInstance() {
        return instance;
    }
}